File: hospitality_core/hospitality_core/page/availability_tool/availability_tool.py
import frappe

@frappe.whitelist()
def check_availability_counts(start_date, end_date):
    # 1. Get all Enabled Rooms
    rooms = frappe.get_all("Hotel Room", 
        fields=["name", "room_type", "status as current_status"],
        filters={"is_enabled": 1},
        order_by="room_number"
    )

    # 2. Get all Overlapping Reservations
    reservations = frappe.db.sql("""
        SELECT room, status, name, guest
        FROM `tabHotel Reservation`
        WHERE status IN ('Reserved', 'Checked In')
        AND arrival_date < %s
        AND departure_date > %s
    """, (end_date, start_date), as_dict=True)

    reservation_map = {r.room: r for r in reservations}

    room_details = []
    summary_map = {}

    for room in rooms:
        res = reservation_map.get(room.name)
        status = "Available"
        details = ""

        if room.current_status == "Out of Order":
            status = "Out of Order"
            details = "Maintenance"
        elif res:
            if res.status == "Checked In":
                status = "Occupied"
            else:
                status = "Reserved"
            details = f"{res.name} ({res.guest})"
        
        room_details.append({
            "room": room.name,
            "room_type": room.room_type,
            "status": status,
            "details": details
        })

        # Summary Logic
        if room.room_type not in summary_map:
            summary_map[room.room_type] = {"room_type": room.room_type, "total": 0, "occupied": 0, "available": 0}
        
        summary_map[room.room_type]["total"] += 1
        if status in ["Occupied", "Reserved", "Out of Order"]:
            summary_map[room.room_type]["occupied"] += 1
        else:
            summary_map[room.room_type]["available"] += 1

    return {
        "summary": list(summary_map.values()),
        "room_details": room_details
    }

File: hospitality_core/hospitality_core/page/availability_tool/availability_tool.js
frappe.pages['availability-tool'].on_page_load = function (wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Room Availability Query',
        single_column: true
    });

    page.add_field({
        fieldname: 'date_range',
        label: 'Date Range',
        fieldtype: 'DateRange',
        default: [frappe.datetime.now_date(), frappe.datetime.add_days(frappe.datetime.now_date(), 1)],
        reqd: 1
    });

    page.set_primary_action('Check', function () {
        let dates = page.fields_dict.date_range.get_value();
        if (!dates) return;

        // DateRange field returns string "YYYY-MM-DD to YYYY-MM-DD" or array depending on version.
        // We assume Array for modern Frappe.
        let start = dates[0];
        let end = dates[1];

        frappe.call({
            method: "hospitality_core.hospitality_core.page.availability_tool.availability_tool.check_availability_counts",
            args: { start_date: start, end_date: end },
            freeze: true,
            callback: function (r) {
                render_results(wrapper, r.message);
            }
        });
    });

    $(wrapper).find('.layout-main-section').append('<div id="avail-results" style="margin-top:20px;"></div>');
}

function render_results(wrapper, data) {
    let summary_html = `<h4>Room Type Summary</h4>
        <table class="table table-bordered">
        <thead><tr style="background-color:#f0f4f7;"><th>Room Type</th><th>Total Rooms</th><th>Unavailable</th><th>Available</th></tr></thead>
        <tbody>`;

    data.summary.forEach(row => {
        let color = row.available > 0 ? 'green' : 'red';
        summary_html += `
            <tr>
                <td><b>${row.room_type}</b></td>
                <td>${row.total}</td>
                <td>${row.occupied}</td>
                <td style="color:${color}; font-weight:bold; font-size:1.2em;">${row.available}</td>
            </tr>
        `;
    });
    summary_html += `</tbody></table>`;

    let details_html = `<h4 style="margin-top:30px;">Detailed Room Status</h4>
        <table class="table table-bordered">
        <thead><tr style="background-color:#f0f4f7;"><th>Room</th><th>Type</th><th>Status</th><th>Reason/Reservation</th></tr></thead>
        <tbody>`;

    data.room_details.forEach(row => {
        let status_color = 'green';
        if (row.status === 'Occupied') status_color = '#d35400';
        if (row.status === 'Reserved') status_color = '#2980b9';
        if (row.status === 'Out of Order') status_color = '#c0392b';

        details_html += `
            <tr>
                <td>${row.room}</td>
                <td>${row.room_type}</td>
                <td><span class="label" style="background-color:${status_color}; color:white; padding: 2px 8px; border-radius: 4px;">${row.status}</span></td>
                <td><small>${row.details || '-'}</small></td>
            </tr>
        `;
    });
    details_html += `</tbody></table>`;

    $('#avail-results').html(summary_html + details_html);
}